### PR TITLE
Added initial implementation of numpy equivalent for trim_zeros to jax

### PR DESCRIPTION
**Pull request by gsp-27**
Hi,

This is my initial attempt to add ```trim_zeros``` function from ```numpy``` to ```jax```.

One problem I am facing is when I add ```_CompileAndCheck``` to the test function, the test fails. I am not sure why this is the case and any help in this regard is appreciated.

Thanks

### CODE REVIEW

**Comment by mattjj** [comment]
Hey @gsp-27 , thanks for the contribution!

It turns out that `trim_zeros` isn't a good candidate for JIT compilation with XLA, because the output shape depends on the values of the input. Because XLA programs have static shapes in their type system, that means we'd need to create and compile a separate program for most input values. That's why you're seeing issues with `_CompileAndCheck`.

We could still include an implementation of `trim_zeros` in `jax.numpy` and just allow it to error when people try to include it in `jit` functions, but that means the upside here is limited so it might also make sense just to leave it out. If we did want to include it, we'd likely need another implementation technique rather than the Python loop implementation here, perhaps based on `lax.while_loop` (though I'm not sure about that).

Overall I'd recommend not adding `trim_zeros` for the above reason.

What do you think?

**Comment by jakevdp** [comment]
See also #3499, which stalled during review.

**Comment by gsp-27** [comment]
Hi,
I am very new to this. I am happy with whatever you suggest.

Thanks,
Gaurav Pathak

**Comment by jakevdp** [comment]
I think it's worth implementing this for now, perhaps with a TODO to use masked output if/when that is more fully supported. It would also be worth replacing the internal `_trim_zeros` utility here: https://github.com/google/jax/blob/d8de6b61411179dcd2f63d7639bbcd69b30ac15f/jax/numpy/lax_numpy.py#L2848-L2852

**Comment by gsp-27** [comment]
Hi Jake,
I looked at #3499 
I maybe wrong here, but the suggested use of ```enumerate``` creates a problem for the case when array is only ```[0]```, in this case numpy returns empty array whereas the jax function will return ```[0]```.

**Comment by jakevdp** [comment]
Ah, you're right.

It would be nice if we could vectorize this operation somehow, though, to avoid manual python loops over single-item device access which can be much more costly for JAX than for numpy, where this implementation comes from. What about something like this?
```
def trim_zeros(filt, trim='fb'):
  nz = asarray(filt) == 0
  if all(nz):    
    return empty(0, filt.dtype)
  start = argmin(nz) if 'f' in trim.lower() else 0
  end = argmin(nz[::-1]) if 'b' in trim.lower() else 0
  return filt[start:len(filt) - end]
```
The downside is that it is `O[N]` even in the best case, where the numpy implementation is `O[1]` when there are no zeros to trim. But the `O[1]` element access is quite slow in JAX, so I thin the `O[N]` vectorized algorithm might be better in practice.

What do you think?

**Comment by gsp-27** [comment]
Hi Jake, 

I will try this out.

Thanks

**Comment by gsp-27** [comment]
Hi,

@jakevdp does the above code look good or did I make some mistake?

Thanks,
Gaurav Pathak

File: tests/lax_numpy_test.py [hunk 1]
```diff
@@ -1227,6 +1227,20 @@ def testPiecewise(self, shape, dtype, ncond, nfunc):
...
 
 
+  @parameterized.named_parameters(jtu.cases_from_list(
+    {"testcase_name": "a_shape={}".format(
+      jtu.format_shape_dtype_string(a_shape, dtype)),
+     "dtype": dtype, "a_shape": a_shape}
+    for dtype in default_dtypes
+    for a_shape in one_dim_array_shapes))
+  def testTrimZeros(self, a_shape, dtype):
+    rng = jtu.rand_default(self.rng())
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1237]: Rather than `jtu.rand_default`, consider `jtu.rand_some_zero` here to increase the chances of having actual zeros to trim!

File: tests/lax_numpy_test.py [hunk 2]
```diff
@@ -1227,6 +1227,20 @@ def testPiecewise(self, shape, dtype, ncond, nfunc):
...
+    np_fun = lambda arg1: np.trim_zeros(arg1)
+    jnp_fun = lambda arg1: jnp.trim_zeros(arg1)
+    args_maker = lambda: [rng(a_shape, dtype)]
+    self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker, check_dtypes=True)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1241]: The test should also cover different options for the `trim` parameter

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks great! A couple comments on the test.

Also, please add `trim_zeros` to the numpy package autosummary list [here](https://github.com/google/jax/blob/master/docs/jax.numpy.rst#jaxnumpy-package) so that it will be part of the HTML documentation.

File: tests/lax_numpy_test.py (hunk 2 above)
Comment (gsp-27) [reply tests/lax_numpy_test.py:1248]: Hi,

I have added the tests to cover all modes of the ```trim``` argument, but I just created different functions to achieve this. Is there a better way to do this using ```args_maker```.

Comment (jakevdp) [reply tests/lax_numpy_test.py:1248]: I think this would be cleaner to do via parametrization. See, for example, how the extra arguments are handled in `testUnique`: https://github.com/google/jax/blob/29aa9bfc8f5dbdf635e877395dc2082cefc8bc87/tests/lax_numpy_test.py#L1460-L1477

The benefit of the parametrized approach is that if one of the test cases fails, the other can still run. It gives a fuller picture of test coverage & results.

Comment (jakevdp) [reply tests/lax_numpy_test.py:1248]: We should test at least `"f"`, `"b"`, and `"fb"`, and maybe even an empty `trim` string.

Comment (gsp-27) [reply tests/lax_numpy_test.py:1242]: Hi,

I have attempted to add the parameterization for the trim argument. I have not included the empty ```trim``` argument test, as by default in ```numpy``` we have the value of ```trim``` set to ```fb```, which I have tried to maintain. Does this sound good?

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!
